### ISSUE TITLE
Why is the log4j dependency included by default when using the library?

### ISSUE THREAD

**Issue by sdutry** [struts-community-plugins/struts2-bootstrap#14]
I see that the logging library log4j is included by default when using the struts2-bootstrap library.

When making an application we generaly tend to choose our own logging implementation, using the SLF4J-bridges to send all logging over the chosen logging implementation.

This means that we currently have to exclude the log4j dependency when using this library.
(It still works when we don't, it just complains when starting the application)

I noticed that the project won't compile when the log4j dependency isn't present. I would like to understand why that is (seeing as there are no class files actualy using any logging).

Would it be possible to make this dependency optional?

**Comment by flofourcade** [comment]
It s possible. I ve done it in a fork by replacing log4j artifact by log4j api. Then, the client app have to provide an impl (native log4j or slf4j,... adapters).

**Comment by sdutry** [comment]
@flofourcade 

I was just surprised that it even needs a logging library, given that there's no reference in any of the java files.

The log4j-api (log4j2) already used by the struts library would be a great choice.

I just tried switching the log4j dependency with the log4j-api (2.6.2) version, but i'm not able to compile the project that way.

`mvn clean compile`
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.3:compile (default-compile) on project struts2-bootstrap-plugin: Fatal error compiling: java.lang.NoClassDefFoundError: org/apache/log4j/Priority: org.apache.log4j.Priority -> [Help 1]
```

**Comment by flofourcade** [comment]
Seems we have a class somewhere which depends on the impl. I ll take a look tomorrow.  We have to fix this issue

**Comment by flofourcade** [comment]
Putting the log4j dependency in "provided" scope make the bootstrap project compile. Strangely, we can see the log4j API initialize itself during the maven compile phase. Very interesting, and funny, I'll investigate

**Comment by sdutry** [comment]
the stacktrace from running it with -X option:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.3:compile (default-compile) on project s
truts2-bootstrap-plugin: Fatal error compiling: java.lang.NoClassDefFoundError: org/apache/log4j/Priority: org.apache.lo
g4j.Priority -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-compiler-p
lugin:3.3:compile (default-compile) on project struts2-bootstrap-plugin: Fatal error compiling
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:216)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.
java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:120)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:347)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:154)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:213)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:157)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:483)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: Fatal error compiling
        at org.apache.maven.plugin.compiler.AbstractCompilerMojo.execute(AbstractCompilerMojo.java:836)
        at org.apache.maven.plugin.compiler.CompilerMojo.execute(CompilerMojo.java:129)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        ... 19 more
Caused by: org.codehaus.plexus.compiler.CompilerException: java.lang.NoClassDefFoundError: org/apache/log4j/Priority
        at org.codehaus.plexus.compiler.javac.JavaxToolsCompiler.compileInProcess(JavaxToolsCompiler.java:172)
        at org.codehaus.plexus.compiler.javac.JavacCompiler.performCompile(JavacCompiler.java:169)
        at org.apache.maven.plugin.compiler.AbstractCompilerMojo.execute(AbstractCompilerMojo.java:825)
        ... 22 more
Caused by: java.lang.RuntimeException: java.lang.NoClassDefFoundError: org/apache/log4j/Priority
        at com.sun.tools.javac.main.Main.compile(Main.java:553)
        at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)
        at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)
        at org.codehaus.plexus.compiler.javac.JavaxToolsCompiler.compileInProcess(JavaxToolsCompiler.java:125)
        ... 24 more
Caused by: java.lang.NoClassDefFoundError: org/apache/log4j/Priority
        at freemarker.log.Log4JLoggerFactory.getLogger(Log4JLoggerFactory.java:65)
        at freemarker.log.Logger.getLogger(Logger.java:255)
        at freemarker.template.utility.SecurityUtilities.<clinit>(SecurityUtilities.java:67)
        at freemarker.ext.beans.BeansWrapper.<clinit>(BeansWrapper.java:146)
        at freemarker.template.ObjectWrapper.<clinit>(ObjectWrapper.java:69)
        at freemarker.core.Configurable.<init>(Configurable.java:132)
        at freemarker.template.Configuration.<init>(Configuration.java:109)
        at freemarker.template.Configuration.<clinit>(Configuration.java:96)
        at org.apache.struts.annotations.taglib.apt.TagAnnotationProcessor.saveTemplates(TagAnnotationProcessor.java:247
)
        at org.apache.struts.annotations.taglib.apt.TagAnnotationProcessor.process(TagAnnotationProcessor.java:160)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:794)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java
:705)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1035)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1176)
        at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
        at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
        at com.sun.tools.javac.main.Main.compile(Main.java:523)
        ... 27 more
Caused by: java.lang.ClassNotFoundException: org.apache.log4j.Priority
        at java.net.URLClassLoader$1.run(URLClassLoader.java:372)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:361)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.net.URLClassLoader.findClass(URLClassLoader.java:360)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 45 more
[ERROR]
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```
it's during the processing of the struts tag annotation. 
There it goes into code from the freemarker library.
It then checks a property using the SecurityUtils, in which a logger is fetched.
Because there is no LoggerFactory a LoggerFactory is created with the Value `LIBRARY_AUTO`. This returns a Log4JFactory

**Comment by flofourcade** [comment]
There's not many solution for us, since the problem is in the freemarker lib. I looked at struts2-core POM and to solve this issue, they added `<optional>true</optional>` to the maven dependency definition in order to make log4j/slf4J dependency not transitive. We could use this solution too.

**Comment by sdutry** [comment]
according to the freemarker documentation, this should get resolved starting from the upcoming `2.4.x` versions

see: [http://freemarker.org/docs/pgui_misc_logging.html](http://freemarker.org/docs/pgui_misc_logging.html)

**Comment by jogep** [comment]
For some reasons the log4j is necessary during compiling. Right now I don't now why. However it i now optional and I also migrated everything else to log4j2.
